lib/interactive_key.rb

```diff
@@ -0,0 +1,553 @@
+            if o.character_state_id
+              d_value[:state_ids][o.character_state_id.to_s] = {} if d_value[:state_ids][o.character_state_id.to_s].nil?
+              d_value[:state_ids][o.character_state_id.to_s][:rows] = {} if d_value[:state_ids][o.character_state_id.to_s][:rows].nil? ## rows which this state identifies
+              d_value[:state_ids][o.character_state_id.to_s][:rows][ @row_hash[otu_collection_object][:object_at_rank] ] = true
+              if @selected_descriptors_hash[d_key] && @selected_descriptors_hash[d_key].include?(o.character_state_id.to_s)
+                d_value[:state_ids][o.character_state_id.to_s][:status] = 'used' ## 'used', 'useful', 'useless'
+              else
+                d_value[:state_ids][o.character_state_id.to_s][:status] = 'useful' ## 'used', 'useful', 'useless'
+              end
+            end
+            unless o.presence.nil?
+              d_value[:state_ids][o.presence.to_s] = {} if d_value[:state_ids][o.presence.to_s].nil?
+              d_value[:state_ids][o.presence.to_s][:rows] = {} if d_value[:state_ids][o.presence.to_s][:rows].nil? ## rows which this state identifies
+              d_value[:state_ids][o.presence.to_s][:rows][ @row_hash[otu_collection_object][:object_at_rank] ] = true
+              if @selected_descriptors_hash[d_key] && @selected_descriptors_hash[d_key].include?(o.presence.to_s)
+                d_value[:state_ids][o.presence.to_s][:status] = 'used' ## 'used', 'useful', 'useless'
+              else
+                d_value[:state_ids][o.presence.to_s][:status] = 'useful' ## 'used', 'useful', 'useless'
+              end
+            end
+            unless o.continuous_value.nil?
+              d_value[:state_ids][o.id] = true
+            end
+            unless o.sample_min.nil?
+              d_value[:state_ids][o.id] = {o_min: o.sample_min.to_f, o_max: o.sample_max.to_f}
+            end
+            taxa_with_unknown_character_states[ @row_hash[otu_collection_object][:object_at_rank] ] = false if @eliminate_unknown == false
+          end
+        end
+      end
+      number_of_taxa_with_unknown_character_states = 0
+      number_of_taxa_with_unknown_character_states = taxa_with_unknown_character_states.select{|key, value| value == true}.count if @eliminate_unknown == false
+
+      descriptor = {}
+      descriptor[:id] = d_key
+      descriptor[:type] = d_value[:descriptor].type
+      descriptor[:name] = d_value[:descriptor].target_name(:key, language)
+      descriptor[:weight] = d_value[:descriptor].weight
+      descriptor[:position] = d_value[:descriptor].position
+      descriptor[:usefulness] = 0
+      descriptor[:status] = d_value[:status] == 'used' ? 'used' : 'useless'
+      descriptor[:description] = d_value[:descriptor].description
+      descriptor[:depiction_ids] = d_value[:descriptor].depictions.order(:position).pluck(:id)
+
+      s = 0
+      case d_value[:descriptor].type
+      when 'Descriptor::Qualitative'
```

Comment:
Seems you could make individual assignment methods for each time, rather than assign out inline?